tensorflow/core/common_runtime/dml/dml_execution_context.cc

```diff
@@ -372,4 +421,176 @@ void DmlExecutionContextImpl::CloseCommandListAndExecute() {
+  // longer than this function call, so make a copy and transfer ownership to
+  // the lambda.
+  absl::InlinedVector<D3D12_RESOURCE_BARRIER, 4> barriers_copy;
+  shared_state_->WriteBatch().emplace_back(
+      [=, barriers = std::move(barriers_copy)]() {
+        shared_state_->impl->ResourceBarrier(barriers);
+      });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+DmlGpuEvent DmlExecutionContext::UavBarrier() {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  shared_state_->WriteBatch().emplace_back(
+      [=]() { shared_state_->impl->UavBarrier(); });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+StatusOr<DmlGpuEvent> DmlExecutionContext::Flush() {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  auto event = shared_state_->next_flush_event;
+  if (shared_state_->WriteBatch().empty()) {
+    --event.fence_value;
+  }
+
+  shared_state_->flush_requested = true;
+  shared_state_->new_function_enqueued.notify_all();
+  return event;
+}
+
+DmlGpuEvent DmlExecutionContext::GetCurrentCompletionEvent() {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+  auto event = shared_state_->next_flush_event;
+
+  if (shared_state_->WriteBatch().empty()) {
+    --event.fence_value;
+  }
+
+  return event;
+}
+
+/*static*/ void DmlExecutionContext::ThreadProc(
+    std::shared_ptr<SharedState> state) {
+  auto last_flush_time = std::chrono::high_resolution_clock::now();
+
+  while (true) {
+    std::chrono::duration<double> elapsed =
+        std::chrono::high_resolution_clock::now() - last_flush_time;
+    auto elapsed_us = elapsed.count() * 1e6;
+
+    std::unique_lock<std::mutex> lock(state->mutex);
+    if (state->exit_requested) {
+      break;
+    }
+
+    auto& batch = state->WriteBatch();
+
+    if (batch.empty()) {
+      // Wait for new work to be batched.
+      state->new_function_enqueued.wait(lock);
+      continue;
+    }
+
+    // Check if it's time to swap the write/execute batches and flush work to
+    // the GPU: this occurs if a flush is explicitly requested, the batch has
+    // reached a certain size, or enough time has elapsed since the last flush.
+    // The goal here is to balance feeding the GPU work while the CPU is
+    // processing more commands and avoiding many small packets.
+    bool flush = false;
+    if (state->flush_requested || batch.size() >= state->batch_flush_size ||
+        elapsed_us >= state->batch_flush_time_us) {
+      state->write_batch_index = (state->write_batch_index + 1) % 2;
+      flush = true;
+      ++state->next_flush_event.fence_value;
+    }
+    state->flush_requested = false;
+
+    // Unlock to allow kernels to resume writing to the new write batch.
+    lock.unlock();
+
+    // Invoke the batched functions and submit the work to the GPU.
+    if (flush) {
+      for (auto& f : batch) {
+        f();
```

Comment:
Is there any particular reason not to execute these as you go along, rather than waiting until the very end when you want to flush? Would that eliminate the need to double-buffer these batches?

Reply:
You mean triple buffering, essentially (batch for kernel threads, batch for execution thread, and batched commands within the command list)? Given that background thread can already become saturated as-is I was thinking to have multiple threads do the recording into the command list; this is a little easier if we batch up the functions and then distribute the recording work. It would be possible with a thread pool by round-robin distributing the function calls, but then we'd need a lock on a command list as well.

Reply:
Going to WontFix this for now, since I'm looking at further splitting up the batch with multiple threads as mentioned.